prow/spyglass/spyglass.go

```diff
@@ -0,0 +1,231 @@
+		pj := lenses[j].Priority
+		if pi == pj {
+			return iname < jname
+		}
+		return pi < pj
+	})
+	return lenses
+}
+
+// Refresh reloads the html view for a given view and a set of objects
+func (s *Spyglass) Refresh(src string, podName string, sizeLimit int64, viewReq *ViewRequest) Lens {
+	lens, ok := s.Lenses[viewReq.ViewName]
+	if !ok {
+		logrus.Errorf("Could not find Lens with name %s.", viewReq.ViewName)
+		return Lens{}
+	}
+	artifacts, err := s.FetchArtifacts(src, podName, sizeLimit, viewReq.ViewerCache[viewReq.ViewName])
+	if err != nil {
+		logrus.WithError(err).Error("Error while fetching artifacts.")
+	}
+
+	view, err := viewers.View(viewReq.ViewName, artifacts, viewReq.ViewData)
+	if err != nil {
+		logrus.WithError(err).Error("Could not find a valid artifact viewer.")
+		return Lens{}
+	}
+	lens.HTMLView = view
+	return lens
+}
+
+// ListArtifacts handles muxing artifact sources to the correct fetcher implementations to list
+// all available artifact names
+func (s *Spyglass) ListArtifacts(src string) ([]string, error) {
+	foundArtifacts := []string{}
+	for _, ep := range s.Eyepieces {
+		jobSource, err := ep.CreateJobSource(src)
+		if err != nil {
+			if err == ErrCannotParseSource {
+				continue
+			}
+			logrus.WithError(err).Error("Error creating job source.")
+		}
+
+		foundArtifacts = append(foundArtifacts, ep.Artifacts(jobSource)...)
+	}
+	foundArtifacts = append(foundArtifacts, neededArtifacts(foundArtifacts)...)
+	return foundArtifacts, nil
+}
+
+func neededArtifacts(haveArtifacts []string) []string {
+	needed := guaranteedArtifacts
+	for _, name := range haveArtifacts {
+		for i := 0; i < len(needed); i++ {
+			if name == needed[i] {
+				needed = append(needed[:i], needed[i+1:]...)
+				break
+			}
+		}
+	}
+	return needed
+}
+
+// FetchArtifacts constructs and returns Artifact objects for each artifact name in the list.
+// This includes getting any handles needed for read write operations, direct artifact links, etc.
+func (s *Spyglass) FetchArtifacts(src string, podName string, sizeLimit int64, artifactNames []string) ([]viewers.Artifact, error) {
+	foundArtifacts := []viewers.Artifact{}
+	foundArtifactNames := []string{}
+	for _, ep := range s.Eyepieces {
+		jobSource, err := ep.CreateJobSource(src)
+		if err != nil {
+			if err == ErrCannotParseSource {
+				continue
+			}
+			logrus.WithError(err).Error("Error creating job source.")
+		}
+		artStart := time.Now()
+		for _, name := range artifactNames {
+			artifact := ep.Artifact(jobSource, name, sizeLimit)
+			foundArtifacts = append(foundArtifacts, artifact)
+			foundArtifactNames = append(foundArtifactNames, artifact.JobPath())
+		}
+		artElapsed := time.Since(artStart)
+		logrus.Info("Retrieved artifacts in ", artElapsed)
+
+	}
+
+	// Special-casing the fetching of in-cluster artifacts that havent been found yet
+	neededArtifactNames := neededArtifacts(foundArtifactNames)
+	for _, ga := range neededArtifactNames {
+		switch ga {
+		case "build-log.txt":
```

Comment:
This seems like a super tenuous workaround for the design and seems fragile. Can we do this more explicitly? If one of our most common use-cases is a special case, doesn't sound like the core design is handling it well.

Reply:
This is meant to special case artifacts that can be found in-cluster and also are not found in GCS. At the moment, this means only the pod logs. Because Prow knows about these artifacts, I think it is safe to special case them.

Reply:
this is a well known artifact that all jobs should produce. abstracting over the sources should simplify some viewer logic... just display build-log, regardless of if it came from the pod directly or not.